File: packages/proxy-hunter-python/proxy_hunter/Proxy.py
import json
from typing import Any, Dict, List, Optional


class Proxy:
    """
    Proxy table data class
    """

    def __init__(
        self,
        proxy: str,
        id: Optional[int] = None,
        latency: Optional[str] = None,
        type: Optional[str] = None,  # Renamed from `type`
        region: Optional[str] = None,
        city: Optional[str] = None,
        country: Optional[str] = None,
        last_check: Optional[str] = None,
        anonymity: Optional[str] = None,
        status: Optional[str] = None,
        timezone: Optional[str] = None,
        longitude: Optional[str] = None,
        private: Optional[str] = None,
        latitude: Optional[str] = None,
        lang: Optional[str] = None,
        useragent: Optional[str] = None,
        webgl_vendor: Optional[str] = None,
        webgl_renderer: Optional[str] = None,
        browser_vendor: Optional[str] = None,
        username: Optional[str] = None,
        password: Optional[str] = None,
        https: Optional[str] = None,  # Optional HTTPS protocol
    ) -> None:
        """
        Proxy constructor.

        :param proxy: Proxy IP address or hostname.
        :param id: Proxy ID.
        :param latency: Proxy latency.
        :param type: Proxy type.
        :param region: Proxy region.
        :param city: Proxy city.
        :param country: Proxy country.
        :param last_check: Last check timestamp.
        :param anonymity: Proxy anonymity.
        :param status: Proxy status.
        :param timezone: Proxy timezone.
        :param longitude: Proxy longitude.
        :param private: Whether the proxy is private.
        :param latitude: Proxy latitude.
        :param lang: Language.
        :param useragent: User agent.
        :param webgl_vendor: WebGL vendor.
        :param webgl_renderer: WebGL renderer.
        :param browser_vendor: Browser vendor.
        :param username: Username for proxy authentication.
        :param password: Password for proxy authentication.
        :param https: HTTPS protocol support.
        """
        self.id = id
        self.proxy = proxy
        self.latency = latency
        self.type = type
        self.region = region
        self.city = city
        self.country = country
        self.last_check = last_check
        self.anonymity = anonymity
        self.status = status
        self.timezone = timezone
        self.longitude = longitude
        self.private = private
        self.latitude = latitude
        self.lang = lang
        self.useragent = useragent
        self.webgl_vendor = webgl_vendor
        self.webgl_renderer = webgl_renderer
        self.browser_vendor = browser_vendor
        self.username = username
        self.password = password
        self.https = https

    def has_credentials(self):
        return self.username and self.password

    def __str__(self):
        attributes = ", ".join(f"{key}: {value}" for key, value in vars(self).items())
        return f"Proxy({attributes})"

    def __repr__(self):
        attributes = ", ".join(f"{key}: {value}" for key, value in vars(self).items())
        return f"Proxy({attributes})"

    def format(self) -> str:
        """
        Format the Person instance into a connection string.

        Returns:
        str: The formatted connection string.
        """
        raw = f"{self.proxy}"
        if self.username is not None and self.password is not None:
            if self.username.strip() and self.password.strip():
                raw = f"{self.proxy}@{self.username}:{self.password}"
        return raw

    def from_dict(self, **kwargs: Any) -> "Proxy":
        """
        Proxy class constructor.

        Args:
            **kwargs: Keyword arguments representing proxy attributes.
        Example:
            print(Proxy("ip:port").from_dict(**dictionary))
        """
        for key, value in kwargs.items():
            setattr(self, key, value)
        return self

    def to_dict(self):
        """
        Transform current result into dict ProxyDB
        """
        properties = {}
        for attr, value in vars(self).items():
            properties[attr] = value
        return properties

    def to_json(self) -> str:
        """
        Convert Proxy instance to JSON string.

        Returns:
        str: JSON representation of the Proxy instance.
        """
        return json.dumps(self.to_dict())


def dict_to_proxy_list(dict_list: List[Dict[str, Any]]) -> List[Proxy]:
    """
    Convert a list of dictionaries to a list of Proxy objects.

    :param dict_list: List of dictionaries, each representing a proxy with its attributes.
    :return: List of Proxy objects instantiated from the provided dictionaries.
    """
    proxy_list = []
    for item in dict_list:
        proxy = Proxy(
            proxy=str(item.get("proxy")),
            id=item.get("id"),
            latency=item.get("latency"),
            type=item.get("type"),
            region=item.get("region"),
            city=item.get("city"),
            country=item.get("country"),
            last_check=item.get("last_check"),
            anonymity=item.get("anonymity"),
            status=item.get("status"),
            timezone=item.get("timezone"),
            longitude=item.get("longitude"),
            private=item.get("private"),
            latitude=item.get("latitude"),
            lang=item.get("lang"),
            useragent=item.get("useragent"),
            webgl_vendor=item.get("webgl_vendor"),
            webgl_renderer=item.get("webgl_renderer"),
            browser_vendor=item.get("browser_vendor"),
            username=item.get("username"),
            password=item.get("password"),
        )
        proxy_list.append(proxy)
    return proxy_list

File: src/PhpProxyHunter/Proxy.php
<?php

namespace PhpProxyHunter;

/**
 * Proxy table data class
 */
class Proxy {
  /** @var int|null */
  public $id = null;

  /** @var string */
  public $proxy;

  /** @var string|null */
  public $latency = null;

  /** @var string|null */
  public $type = null;

  /** @var string|null */
  public $region = null;

  /** @var string|null */
  public $city = null;

  /** @var string|null */
  public $country = null;

  /** @var string|null */
  public $last_check = null;

  /** @var string|null */
  public $anonymity = null;

  /** @var string|null */
  public $status = null;

  /** @var string|null */
  public $timezone = null;

  /** @var string|null */
  public $longitude = null;

  /** @var string|null */
  public $private = null;

  /** @var string|null */
  public $latitude = null;

  /** @var string|null */
  public $lang = null;

  /** @var string|null */
  public $useragent = null;

  /** @var string|null */
  public $webgl_vendor = null;

  /** @var string|null */
  public $webgl_renderer = null;

  /** @var string|null */
  public $browser_vendor = null;

  /** @var string|null */
  public $username = null;

  /** @var string|null */
  public $password = null;

  /** @var string|null */
  public $https = 'false';

  /**
   * Creates a Proxy object from either a proxy string or an associative array.
   *
   * If $proxy is a string, it becomes the proxy address directly.
   * If $proxy is an array, all matching keys will be mapped to the object's properties.
   *
   * @param string|array $proxy          Proxy string (e.g. "1.2.3.4:8080") or array with proxy data.
   * @param string|null  $latency        Proxy latency in ms.
   * @param string|null  $type           Proxy type (HTTP, SOCKS4, SOCKS5, etc.).
   * @param string|null  $region         Region of the proxy.
   * @param string|null  $city           City of the proxy.
   * @param string|null  $country        Country of the proxy.
   * @param string|null  $last_check     Timestamp of the last check.
   * @param string|null  $anonymity      Level of anonymity.
   * @param string|null  $status         Status (e.g. "online", "offline").
   * @param string|null  $timezone       Proxy timezone.
   * @param string|null  $longitude      Longitude coordinate.
   * @param string|null  $private        Private/public flag.
   * @param string|null  $latitude       Latitude coordinate.
   * @param string|null  $lang           Language code.
   * @param string|null  $useragent      User-Agent string.
   * @param string|null  $webgl_vendor   WebGL vendor fingerprint.
   * @param string|null  $webgl_renderer WebGL renderer fingerprint.
   * @param string|null  $browser_vendor Browser vendor fingerprint.
   * @param string|null  $username       Username (if proxy has authentication).
   * @param string|null  $password       Password (if proxy has authentication).
   * @param int|null     $id             Internal database ID.
   * @param string|null  $https          Whether proxy supports HTTPS ("true"/"false").
   */
  public function __construct(
    $proxy,
    $latency = null,
    $type = null,
    $region = null,
    $city = null,
    $country = null,
    $last_check = null,
    $anonymity = null,
    $status = null,
    $timezone = null,
    $longitude = null,
    $private = null,
    $latitude = null,
    $lang = null,
    $useragent = null,
    $webgl_vendor = null,
    $webgl_renderer = null,
    $browser_vendor = null,
    $username = null,
    $password = null,
    $id = null,
    $https = 'false'
  ) {
    $this->id             = $id;
    $this->latency        = $latency;
    $this->type           = $type;
    $this->region         = $region;
    $this->city           = $city;
    $this->country        = $country;
    $this->last_check     = $last_check;
    $this->anonymity      = $anonymity;
    $this->status         = $status;
    $this->timezone       = $timezone;
    $this->longitude      = $longitude;
    $this->private        = $private;
    $this->latitude       = $latitude;
    $this->lang           = $lang;
    $this->useragent      = $useragent;
    $this->webgl_vendor   = $webgl_vendor;
    $this->webgl_renderer = $webgl_renderer;
    $this->browser_vendor = $browser_vendor;
    $this->username       = $username;
    $this->password       = $password;
    $this->https          = $https;

    if (is_string($proxy)) {
      $this->proxy = $proxy;
    } elseif (is_array($proxy) && isset($proxy['proxy'])) {
      foreach ($proxy as $key => $value) {
        if (property_exists($this, $key) && $key !== 'proxy') {
          $this->$key = $value;
        }
      }
    } else {
      throw new \InvalidArgumentException('Proxy must be a string or an array with a "proxy" key.');
    }
  }

  /**
   * Returns a JSON representation of the Proxy object.
   *
   * @param bool $pretty   If true, returns pretty-printed JSON.
   * @param bool $notEmpty If true, only includes properties that are not null or empty.
   * @return string        JSON encoded string representing the Proxy object.
   */
  public function toJson($pretty = false, $notEmpty = false) {
    $data = get_object_vars($this);
    if ($notEmpty) {
      $data = array_filter($data, function ($value) {
        return !is_null($value) && $value !== '';
      });
    }
    if ($pretty) {
      return json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
    }
    return json_encode($data, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
  }

  /**
   * Magic string conversion.
   * Allows (string)$proxyObj and "class proxy: $proxyObj" to produce a useful string.
   * We return a compact JSON of non-empty properties for readability.
   *
   * @return string
   */
  public function __toString() {
    // Return in format: proxy@user:pass
    // Omit parts that are empty: if no username, return just proxy; if no password, return proxy@user
    $vars  = get_object_vars($this);
    $proxy = isset($vars['proxy']) ? (string)$vars['proxy'] : '';
    $user  = isset($vars['username']) && $vars['username'] !== null && $vars['username'] !== '' ? (string)$vars['username'] : null;
    $pass  = isset($vars['password']) && $vars['password'] !== null && $vars['password'] !== '' ? (string)$vars['password'] : null;

    if ($proxy === '' && $user === null && $pass === null) {
      return '';
    }

    if ($user === null) {
      return $proxy;
    }

    $out = $proxy . '@' . $user;
    if ($pass !== null) {
      $out .= ':' . $pass;
    }
    return $out;
  }
}
